File: Python/sudoku-solver.py
class Solution:
    # @param {character[][]} board
    # @return {void} Do not return anything, modify board in-place instead.
    def solveSudoku(self, board):
        def is_valid(i, j):
            val = board[i][j]
            board[i][j] = '.'
            for k in xrange(9):
                if(board[i][k]==val or board[k][j]==val or board[(i/3)*3+k/3][(j/3)*3+k%3]==val):
                    board[i][j] = val
                    return False
            board[i][j] = val
            return True
        
            
        def DFS(board):
            for i in xrange(9):
                for j in xrange(9):
                    if board[i][j]=='.':
                        for k in '123456789':
                            board[i][j] = k
                            if is_valid(i, j) and DFS(board):
                                return True
                        board[i][j] = '.'
                        return False
            return True
        DFS(board)

File: C++/sudoku-solver.cpp
class Solution {
private:
    bool is_valid(std::vector<std::vector<char>> &board, int i, int j) {
        int val = board[i][j];
        board[i][j] = '.';
        for(int k=0; k<9; ++k) {
            if(board[i][k]==val || board[k][j]==val || board[(i/3)*3+k/3][(j/3)*3+k%3]==val) {
                board[i][j] = val;
                return false;
            }
        }
        board[i][j] = val;
        return true;
    }
    bool DFS(std::vector<std::vector<char>> &board) {
        std::string numbers = "123456789";
        for(int i=0; i<9; ++i) {
            for(int j=0; j<9; ++j) {
                if(board[i][j]=='.') {
                    for(char k : numbers) {
                        board[i][j] = k;
                        if(is_valid(board, i, j) && DFS(board))
                            return true;
                    }
                    board[i][j] = '.';
                    return false;
                }
            }
        }
        return true;
    }
public:
    void solveSudoku(vector<vector<char>> &board) {
        DFS(board);
    }
};
